### PR TITLE
[batch] create batch efficiently

### PR DESCRIPTION
**Pull request by jigold**
Stacked on #6075 

- Changed the job id to be a compound key (batch_id, job_id) where batch_id is unique while job_id is 1->N
- Added a batch builder to create the batch in the database as efficiently as possible in one transaction (or at least that was my intent)
- Changed both client interfaces to have a `run` function rather than `close`. Sends one request once the batch has been fully created.

### CODE REVIEW

**Comment by danking** [comment]
I'm also a bit of the mind that this is a big change that's mentally keeping me from getting started on other batch changes so maybe we should just merge and work through whatever happens, at least so that development can continue.

File: batch/batch/aioclient.py [hunk 1]
```diff
@@ -2,29 +2,38 @@
...
 
 import hailjwt as hj
 
 from .requests_helper import filter_params
-from . import schemas
 
 
 class Job:
-    def __init__(self, client, id, attributes=None, parent_ids=None, _status=None):
+    def __init__(self, client, batch_id, job_id, attributes=None, parent_ids=None, _status=None):
```
[the remark this answers is not in the document]
Comment (jigold) [reply batch/batch/aioclient.py]: The issue was feeding back the job id back to the Python object on the client side. Can I guarantee the order of records inserted into the database is the same as the order of jobs submitted by the user? If not, this is the only way to guarantee the job id is correct.

File: batch/batch/aioclient.py [hunk 2]
```diff
@@ -49,22 +67,54 @@ def __init__(self, client, id, attributes=None, parent_ids=None, _status=None):
...
-        return await self.client._get('/jobs/{}/log'.format(self.id))
+        if self._batch_is_open:
+            raise ValueError("Job is not running yet")
+
+        return await self.client._get('/batches/{}/jobs/{}/log'.format(*self.id))
 
 
 class Batch:
-    def __init__(self, client, id, attributes):
+    def __init__(self, client, id, attributes, _doc=None, _is_open=False):
```
[the remark this answers is not in the document]
Comment (jigold) [reply batch/batch/aioclient.py]: ok.

File: batch/batch/server/database.py [hunk 3]
```diff
@@ -1,6 +1,78 @@
...
+        assert self._is_open
+        if len(self._jobs) == 0:
+            self._jobs_sql = self._db.jobs.new_record_template(*items)
+        self._jobs.append(dict(items))
+
+    def create_job_parent(self, **items):
+        assert self._is_open
+        if len(self._jobs_parents) == 0:
+            self._jobs_parents_sql = self._db.jobs_parents.new_record_template(*items)
+        self._jobs_parents.append(dict(items))
```
[the remark this answers is not in the document]
Comment (jigold) [reply batch/batch/server/database.py]: No it was a view and the sql code failed with that.

File: batch/batch/server/database.py [hunk 4]
```diff
@@ -23,43 +95,45 @@ def __init__(self, db):
...
+                          WHERE `{self.name}`.state IN %s AND `{jobs_parents_name}`.batch_id = %s AND `{jobs_parents_name}`.job_id = %s"""
 
-                await cursor.execute(sql, (('Created', 'Ready'), id))
+                await cursor.execute(sql, (('Created', 'Ready'), batch_id, job_id))
                 result = await cursor.fetchall()
-                return [record['id'] for record in result]
+                return [(record['batch_id'], record['job_id']) for record in result]
 
     async def get_record_by_pod(self, pod):
         records = await self.get_records_where({'pod_name': pod})
```
Comment (jigold) [inline batch/batch/server/database.py]: should assert pod is not None

**Comment by jigold** [comment]
I'm going to try and split this PR into 2-3 pieces.

**Comment by jigold** [comment]
I'm closing this in favor of smaller PRs like #6183

**Pull request closed without merging**